ethers-contract/src/multicall.rs

```diff
@@ -0,0 +1,204 @@
+extern crate hex;
+
+use ethers_core::{
+    abi::{Detokenize, Function, Token},
+    types::{Address, BlockNumber, NameOrAddress, TransactionRequest, U256},
+};
+use ethers_providers::JsonRpcClient;
+use ethers_signers::{Client, Signer};
+use hex::FromHex;
+use lazy_static::lazy_static;
+
+use std::{collections::HashMap, sync::Arc};
+
+use crate::call::{ContractCall, ContractError};
+use crate::multicall_contract::MulticallContract;
+
+lazy_static! {
+    static ref ADDRESS_BOOK: HashMap<U256, Address> = {
+        let mut m = HashMap::new();
+
+        // mainnet
+        let addr = <[u8; 20]>::from_hex("eefba1e63905ef1d7acba5a8513c70307c1ce441").expect("Decoding failed");
+        m.insert(U256::from(1u8), Address::from(addr));
+
+        // rinkeby
+        let addr = <[u8; 20]>::from_hex("42ad527de7d4e9d9d011ac45b31d8551f8fe9821").expect("Decoding failed");
+        m.insert(U256::from(4u8), Address::from(addr));
+
+        // goerli
+        let addr = <[u8; 20]>::from_hex("77dca2c955b15e9de4dbbcf1246b4b85b651e50e").expect("Decoding failed");
+        m.insert(U256::from(5u8), Address::from(addr));
+
+        // kovan
+        let addr = <[u8; 20]>::from_hex("2cc8688c5f75e365aaeeb4ea8d6a480405a48d2a").expect("Decoding failed");
+        m.insert(U256::from(42u8), Address::from(addr));
+
+        m
+    };
+}
+
+#[derive(Clone)]
+struct Call {
+    target: Address,
+    data: Vec<u8>,
+    function: Function,
+}
+
+pub struct Multicall<P, S> {
+    address: Option<Address>,
+    calls: Vec<Call>,
+    block: Option<BlockNumber>,
+    client: Arc<Client<P, S>>,
+}
+
+impl<P, S> Multicall<P, S> {
+    pub fn new(
```

Comment:
I think instead we could simplify this a bit more and make `fn new(client: ...)`. We'd have `address` and `block` default to `None`, and add builder pattern methods which would set them separately, what do you think? I think that it's more likely that when people actually use the `MultiCall` contract, they'd just use it with the defaults

Reply:
I think this should be the better approach! Thanks. One doubt regarding this:
I would need to have a method `new` as well as `build` in order to support such functionality. Because if I need to instantiate the `Contract` instance for multicall, I would need the address. And if I'm expecting the address to be possibly supplied as an argument by the user, then I should have a `build` function that finally instantiates the contract. What do you think about this?

Reply:
Or I can set the field as: `contract: Option<Contract<P, S>>`. So in that case, it will be `None` if the appropriate Network ID was not found. And we then expect the user to supply contract address. The `build` would then panic if no address was supplied and `contract` is also `None`

Reply:
> So in any case, I'd need to store client?

You can access a contract's `Client` by calling the `client()` method https://docs.rs/ethers-contract/0.1.3/ethers_contract/struct.Contract.html#method.client

Reply:
Yes, but if we don't take the `address` field in the constructor `Multicall::new`, then in the case where the user would be adding it using the builder pattern later, we cannot construct an instance of the Multicall contract, and so we would need to store the client, right?